File: ios/Plugin/DeviceLocale.swift
import Foundation

@objc public class DeviceLocale: NSObject {
    /**
     Retrieves the region code of the current locale.
     
     - Returns: The region code as a string or `nil` if it is not available.
     */
    @objc public func getRegionCode() -> String? {
        if #available(iOS 16, *) {
            return Locale.current.region?.identifier ?? nil
        }
        
        return Locale.current.regionCode
    }
    
    /**
     Retrieves the country name corresponding to the current locale.
     
     - Returns: The country name as a string or `nil` if it is not available.
     */
    @objc public func getCountry() -> String? {
        let currentLocale = Locale.current
        if let countryName = currentLocale.localizedString(forRegionCode: getRegionCode() ?? "") {
            return countryName
        }
        
        return nil
    }
    
    /**
     Retrieves the language code of the current locale.
     
     - Returns: The language code as a string or `nil` if it is not available.
     */
    @objc public func getLanguageCode() -> String? {
        if #available(iOS 16, *) {
            return Locale.current.language.languageCode?.identifier ?? nil
        }
        
        return Locale.current.languageCode
    }
    
    /**
     Retrieves the language corresponding to the current locale.
     
     - Returns: The language as a string or `nil` if it is not available.
     */
    @objc public func getLanguage() -> String? {
        let currentLocale = Locale.current
        if let language = currentLocale.localizedString(forLanguageCode: getLanguageCode() ?? "") {
            return language
        }
        
        return nil
    }
    
    /**
     Retrieves the currency symbol of the current locale.
     
     - Returns: The currency symbol as a string or `nil` if it is not available.
     */
    @objc public func getCurrencySymbol() -> String? {
        return Locale.current.currencySymbol
    }
    
    /**
     Retrieves the currency code of the current locale.
     
     - Returns: The currency code as a string or `nil` if it is not available.
     */
    @objc public func getCurrencyCode() -> String? {
        if #available(iOS 16, *) {
            return Locale.current.currency?.identifier
        } else {
            return Locale.current.currencyCode
        }
    }
    
    /**
     Retrieves the currency name corresponding to the current locale.
     
     - Returns: The currency name as a string or `nil` if it is not available.
     */
    @objc public func getCurrencyName() -> String? {
        let currentLocale = Locale.current
        if let currencyName = currentLocale.localizedString(forCurrencyCode: getCurrencyCode() ?? "") {
            return currencyName
        }
        
        return nil
    }
    
    /**
     Retrieves the time zone identifier of the current locale.
     
     - Returns: The time zone identifier as a string or `nil` if it is not available.
     */
    @objc public func getTimeZone() -> String? {
        return String(TimeZone.current.identifier)
    }
}

File: android/src/main/java/com/sarif/capacitor/device/locale/DeviceLocale.java
package com.sarif.capacitor.device.locale;

import java.util.Currency;
import java.util.Locale;
import java.util.TimeZone;

public class DeviceLocale {
    /**
     * Returns the region code of the current locale.
     *
     * @return The region code.
     */
    public String getRegionCode() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
            return Locale.getDefault(Locale.Category.FORMAT).getCountry();
        }

        return Locale.getDefault().getCountry();
    }

    /**
     * Returns the country name of the current locale.
     *
     * @return The country name.
     */
    public String getCountry() {
        Locale currentLocale = Locale.getDefault();
        String regionCode = getRegionCode();
        if (regionCode != null) {
            String countryName = currentLocale.getDisplayCountry(new Locale("", regionCode));
            return countryName;
        }

        return null;
    }

    /**
     * Returns the language code of the current locale.
     *
     * @return The language code.
     */
    public String getLanguageCode() {
        return Locale.getDefault().getLanguage();
    }

    /**
     * Returns the language of the current locale.
     *
     * @return The language.
     */
    public String getLanguage() {
        Locale currentLocale = Locale.getDefault();
        String languageCode = getLanguageCode();
        if (languageCode != null) {
            String language = currentLocale.getDisplayLanguage(new Locale(languageCode));
            return language;
        }

        return null;
    }

    /**
     * Returns the symbol of the currency used in the current locale.
     *
     * @return The currency symbol.
     */
    public String getCurrencySymbol() {
        return Currency.getInstance(Locale.getDefault()).getSymbol();
    }

    /**
     * Returns the ISO 4217 code of the currency used in the current locale.
     *
     * @return The currency code.
     */
    public String getCurrencyCode() {
        return Currency.getInstance(Locale.getDefault()).getCurrencyCode();
    }

    /**
     * Returns the name of the currency used in the current locale.
     *
     * @return The currency name.
     */
    public String getCurrencyName() {
        Locale currentLocale = Locale.getDefault();
        String currencyCode = getCurrencyCode();
        if (currencyCode != null) {
            String currencyName = currentLocale.getDisplayLanguage(new Locale("", currencyCode));
            return currencyName;
        }

        return null;
    }

    /**
     * Returns the ID of the time zone used in the current locale.
     *
     * @return The time zone ID.
     */
    public String getTimeZone() {
        return TimeZone.getDefault().getID();
    }
}
